pkg/wakaama/contrib/objects/on_off_switch.c

```diff
@@ -0,0 +1,492 @@
+    }
+    return COAP_205_CONTENT;
+}
+
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
+                        lwm2m_data_t ** data_array, lwm2m_object_t * object)
+{
+    (void)context;
+
+    lwm2m_obj_on_off_switch_inst_t *instance;
+    uint8_t result = COAP_404_NOT_FOUND;
+    int i = 0;
+
+    mutex_lock(&_on_off_switch_object.lock);
+
+    /* try to get the requested instance from the object list */
+    instance = (lwm2m_obj_on_off_switch_inst_t *)lwm2m_list_find(object->instanceList, instance_id);
+    if (!instance) {
+        DEBUG("[lwm2m:on_off_switch:read]: can't find instance %" PRId16 "\n", instance_id);
+        result = COAP_404_NOT_FOUND;
+        goto free_out;
+    }
+
+    /* if the number of resources is not specified, we need to read all resources */
+    if (!*num_data) {
+        DEBUG("[lwm2m:on_off_switch:read]: reading all resources\n");
+
+        uint16_t res_list[] = {
+            LWM2M_ON_OFF_SWITCH_DIGITAL_INPUT_STATE_ID,
+            LWM2M_ON_OFF_SWITCH_ON_TIME_ID,
+            LWM2M_ON_OFF_SWITCH_OFF_TIME_ID,
+            LWM2M_ON_OFF_SWITCH_APP_TYPE_ID
+        };
+
+        /* allocate structures to return resources */
+        int res_num = ARRAY_SIZE(res_list);
+        *data_array = lwm2m_data_new(res_num);
+
+        if (NULL == *data_array) {
+            result = COAP_500_INTERNAL_SERVER_ERROR;
+            goto free_out;
+        }
+
+        /* return the number of resources being read */
+        *num_data = res_num;
+
+        /* set the IDs of the resources in the data structures */
+        for (i = 0; i < res_num; i++) {
+            (*data_array)[i].id = res_list[i];
+        }
+    }
+
+    /* now get the values */
+    i = 0;
+    do {
+        DEBUG("[lwm2m:on_off_switch:read]: reading resource %" PRId16 "\n", (*data_array)[i].id);
+        result = _get_value(&(*data_array)[i], instance);
+        i++;
+    } while (i < *num_data && COAP_205_CONTENT == result);
+
+free_out:
+    mutex_unlock(&_on_off_switch_object.lock);
+    return result;
+}
+
+static uint8_t _write_cb(lwm2m_context_t * context, uint16_t instance_id, int num_data,
+                         lwm2m_data_t * data_array, lwm2m_object_t * object, lwm2m_write_type_t write_type)
+{
+    (void)context;
+    (void)write_type;
+
+    lwm2m_obj_on_off_switch_inst_t *instance;
+    uint8_t result = COAP_204_CHANGED;
+
+    mutex_lock(&_on_off_switch_object.lock);
+
+    /* try to get the requested instance from the object list */
+    instance = (lwm2m_obj_on_off_switch_inst_t *)lwm2m_list_find(object->instanceList, instance_id);
+    if (!instance) {
+        DEBUG("[lwm2m:on_off_switch:write]: can't find instance %" PRId16 "\n", instance_id);
+        result = COAP_404_NOT_FOUND;
+        goto free_out;
+    }
+
+    for (int i = 0; i < num_data && result == COAP_204_CHANGED; i++) {
+        switch (data_array[i].id) {
+        case LWM2M_ON_OFF_SWITCH_DIGITAL_INPUT_STATE_ID: {
```

Comment:
I think it would be nice to factor the inner block of the loop into a `_set_value(lwm2m_data_t value,  lwm2m_obj_on_off_switch_inst_t *instance)` function. That way the structure resembles the `_read_cb`, makes for easier code reading? I think? :thinking:

Reply:
Refactored